File: main/couples-holding-hands/couples_holding_hands.rb
# Disjoint set data structure with union by rank and (full) path compression.
class DisjointSets
  attr_reader :joins

  def initialize(size)
    @elems = Array.new(size, -1)
    @joins = 0
  end

  def union(i, j)
    # find the ancestors and stop if they are already the same
    i = find_set(i)
    j = find_set(j)
    return if i == j

    # unite by rank
    if @elems[i] > @elems[j]
      # j has superior (more negative) rank, so it will be the parent
      @elems[i] = j
    else
      # if i and j have the same rank, promote i
      @elems[i] -= 1 if @elems[i] == @elems[j]

      # i has superior (more negative) rank, so it will be the parent
      @elems[j] = i
    end

    @joins += 1
  end

  private def find_set(i)
    # find the ancestor
    j = i
    j = @elems[j] until @elems[j].negative?

    # compress the path
    until i == j
      parent = @elems[i]
      @elems[i] = j
      i = parent
    end

    j
  end
end

# @param {Integer[]} row
# @return {Integer}
def min_swaps_couples(row)
  sets = DisjointSets.new(row.size / 2)
  row.each_slice(2) { |u, v| sets.union(u / 2, v / 2) }
  sets.joins
end

File: main/couples-holding-hands/couples-holding-hands.c
static inline void makeSets(int *const a, const int n)
{
    // set every element to -1
    memset(a, 0xFF, n * sizeof *a);
}

static int findSet(int *const a, int i)
{
    // find the ancestor
    int j = i;
    while (a[j] >= 0) j = a[j];

    // compress the path
    while (i != j) {
        const int next = a[i];
        a[i] = j;
        i = next;
    }

    return j;
}

static int unite(int *const a, int i, int j)
{
    // find the ancestors and stop if they are the same
    i = findSet(a, i);
    j = findSet(a, j);
    if (i == j) return 0;

    // union by rank
    if (a[i] > a[j]) {
        // j has superior (more negative) rank, so make it the parent
        a[i] = j;
    } else {
        // if i and j both have the same rank, promote i
        if (a[i] == a[j]) --a[i];

        // i has superior (more negative) rank, so make it the parent
        a[j] = i;
    }

    return 1;
}

int minSwapsCouples(int *row, const int rowSize)
{
    const int n = rowSize / 2;
    int a[n];
    makeSets(a, n);

    int swaps = 0;
    for (int pos = 0; pos != rowSize; pos += 2)
        swaps += unite(a, row[pos] / 2, row[pos + 1] / 2);

    return swaps;
}
